SmartDeviceLink/SDLSecondaryTransportManager.m

```diff
@@ -682,29 +682,50 @@ - (void)sdl_onAppStateUpdated:(NSNotification *)notification {
-                // Start a background task so we can tear down the TCP socket successfully before the app is suspended
+                SDLLogD(@"Starting background task to keep TCP transport alive");
+                strongSelf.backgroundTaskManager.taskEndedHandler = [strongSelf sdl_backgroundTaskEndedHandler];
                 [strongSelf.backgroundTaskManager startBackgroundTask];
-                [strongSelf.stateMachine transitionToState:SDLSecondaryTransportStateConfigured];
             } else {
-                SDLLogD(@"App will go to background. TCP transport already disconnected: %@", strongSelf.stateMachine.currentState);
+                SDLLogD(@"TCP transport already disconnected");
             }
         } else if (notification.name == UIApplicationDidBecomeActiveNotification) {
-            if ([strongSelf.stateMachine isCurrentState:SDLSecondaryTransportStateConfigured]
+            SDLLogD(@"App entered the foreground");
+            if ([strongSelf.stateMachine isCurrentState:SDLSecondaryTransportStateRegistered]) {
+                SDLLogD(@"TCP transport has not yet been shutdown");
+                [strongSelf.backgroundTaskManager endBackgroundTask];
+            } else if ([strongSelf.stateMachine isCurrentState:SDLSecondaryTransportStateConfigured]
                 && strongSelf.secondaryTransportType == SDLSecondaryTransportTypeTCP
                 && [strongSelf sdl_isTCPReady]
                 && [strongSelf sdl_isHMILevelNonNone]) {
-                SDLLogD(@"Resuming TCP transport because the app came into the foreground");
+                SDLLogD(@"Restarting the TCP transport");
                 [strongSelf.backgroundTaskManager endBackgroundTask];
                 [strongSelf.stateMachine transitionToState:SDLSecondaryTransportStateConnecting];
             } else {
-                SDLLogD(@"App returning to foreground. TCP transport not ready to connect: %@", strongSelf.stateMachine.currentState);
+                SDLLogD(@"TCP transport not ready to start");
             }
         }
     });
 }
 
+/// Handles a notification that the background task has ended. If the app is still in the background, the TCP transport disconnects. If the app has re-entered the foreground or the manager has shutdown then the notification is ignored.
+/// @return A background task ended handler
+- (nullable void (^)(void))sdl_backgroundTaskEndedHandler {
+    __weak typeof(self) weakSelf = self;
+    return ^{
+        __strong typeof(self) strongSelf = weakSelf;
+        if (strongSelf.sdl_getAppState == UIApplicationStateActive) {
+            SDLLogD(@"App has been foregrounded. Ignoring notification that the background task ended.");
+        } else if ([strongSelf.stateMachine isCurrentState:SDLSecondaryTransportStateStopped]) {
+            SDLLogD(@"Manager has been stopped. Ignoring notification that the background task ended.");
```

Comment:
I don't know that these are necessary. I think they can either be verbose logs or removed.

Reply:
I changed to verbose logs